// @llm-tools/embedjs/dist/embeddings/ollama-embeddings.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OllamaEmbeddings = void 0;
const ollama_1 = require("@langchain/community/embeddings/ollama");
class OllamaEmbeddings {
    constructor(options) {
        Object.defineProperty(this, "model", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.model = new ollama_1.OllamaEmbeddings({
            model: options.model,
            baseUrl: options.baseUrl,
            keepAlive: options?.keepAlive,
            headers: options?.headers,
            requestOptions: options?.requestOptions,
        });
    }
    async getDimensions() {
        const sample = await this.model.embedDocuments(['sample']);
        return sample[0].length;
    }
    async embedDocuments(texts) {
        return this.model.embedDocuments(texts);
    }
    async embedQuery(text) {
        return this.model.embedQuery(text);
    }
}
exports.OllamaEmbeddings = OllamaEmbeddings;
