// @llm-tools/embedjs/dist/embeddings/openai-3large-embeddings.js
import { OpenAIEmbeddings } from '@langchain/openai';
export class OpenAi3LargeEmbeddings {
    constructor(options) {
        Object.defineProperty(this, "model", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "dynamicDimension", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.dynamicDimension = options?.dynamicDimension ?? 3072;
        this.model = new OpenAIEmbeddings({
            modelName: 'text-embedding-3-large',
            configuration: options?.configuration,
            maxConcurrency: options?.maxConcurrency ?? 3,
            maxRetries: options?.maxRetries ?? 5,
        });
    }
    async getDimensions() {
        return this.dynamicDimension;
    }
    async embedDocuments(texts) {
        return this.model.embedDocuments(texts);
    }
    async embedQuery(text) {
        return this.model.embedQuery(text);
    }
}
