// @llm-tools/embedjs/dist/embeddings/openai-3large-embeddings.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAi3LargeEmbeddings = void 0;
const openai_1 = require("@langchain/openai");
class OpenAi3LargeEmbeddings {
    constructor(options) {
        Object.defineProperty(this, "model", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "dynamicDimension", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.dynamicDimension = options?.dynamicDimension ?? 3072;
        this.model = new openai_1.OpenAIEmbeddings({
            modelName: 'text-embedding-3-large',
            configuration: options?.configuration,
            maxConcurrency: options?.maxConcurrency ?? 3,
            maxRetries: options?.maxRetries ?? 5,
        });
    }
    async getDimensions() {
        return this.dynamicDimension;
    }
    async embedDocuments(texts) {
        return this.model.embedDocuments(texts);
    }
    async embedQuery(text) {
        return this.model.embedQuery(text);
    }
}
exports.OpenAi3LargeEmbeddings = OpenAi3LargeEmbeddings;
